File: Bag of Tokens.go
func bagOfTokensScore(tokens []int, power int) int {
    n := len(tokens)
    if n == 0 {
        return 0
    }

    slices.Sort(tokens)
    var maxx, score int
    i := 0
    j := n-1
    for i < n && power >= tokens[i] && i <= j {
        for i < n && power >= tokens[i] && i <= j {
            power -= tokens[i]
            i += 1
            score += 1
        }

        maxx = max(maxx, score)
        if score > 0 && i <= j {
            power += tokens[j]
            j -= 1
            score -= 1
        }
    }
    return maxx
}

File: Bag of Tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        int n = tokens.size();
        if (n == 0) return 0;
        sort(tokens.begin(), tokens.end());
        int maxx = 0;
        int score = 0;
        int i = 0;
        int j = n-1;
        while (i < n && power >= tokens.at(i) && i <= j) {
            while (i < n && power >= tokens.at(i) && i <= j) {
                power -= tokens.at(i);
                i += 1;
                score += 1;
            }

            maxx = max(maxx, score);
            if (score > 0 && i <= j) {
                power += tokens.at(j);
                j -= 1;
                score -= 1;
            }
        }
        return maxx;
    }
};
